_posts/2025-05-09-opentelemetry-zero-code-instrumentation.md

```diff
@@ -0,0 +1,537 @@
+        # you can also add attributes or events
+        trace.get_current_span().set_attribute("order.id", order_id)
+        trace.get_current_span().add_event("Order processing started")
+
+        # simulate some work
+        time.sleep(1)
+
+        result = "done"
+        trace.get_current_span().add_event("Order processing finished")
+        return result
+```
+
+In the handler we create a function `process_order` that simulates some custom business logic to process an order. In the function a new span is initialized. The order id is read from a request header named `id` and included in the custom span as an attribute. We also add some events to the span to indicate the processing state.
+
+Update the `stack.yaml` and configure the OpenTelemetry agent for the order function:
+
+```yaml
+functions:
+  order:
+    lang: python3-http
+    handler: ./order
+    image: ${REPO:-ttl.sh}/${OWNER:-openfaas}/order:0.0.1
+    environment:
+      OTEL_SERVICE_NAME: order.${NAMESPACE:-openfaas-fn}
+      OTEL_TRACES_EXPORTER: ${EXPORTER:-otlp}
+      OTEL_METRICS_EXPORTER: none
+      OTEL_LOGS_EXPORTER: none
+      OTEL_EXPORTER_OTLP_ENDPOINT: ${OTEL_EXPORTER_OTLP_ENDPOINT:-alloy.monitoring.svc.cluster.local:4317}
+      OTEL_EXPORTER_OTLP_TRACES_INSECURE: true
+```
+
+Build and deploy the function:
+
+```sh
+faas-cli up --filter order
+```
+
+Invoke the function and inspect the trace in Grafana. You should see the custom `process_order` span as part of the function invocation.
+
+```sh
+curl -i "http://127.0.0.1:8080/function/order?id=1"
+```
+
+![Screenshot of the Grafana Explorer showing an invocation trace with a custom span](/images/2025-05-opentelemetry/manual-span-trace.png)
+> Trace detail showing the function invocation with the custom span. The span has the order id attribute and events that we added to it.
+
+## OpenTelemetry on OpenFaaS Edge
+
+Collecting telemetry is also supported on [OpenFaaS Edge](https://docs.openfaas.com/edge/overview/). The Alloy collector can be deployed as an additional service. Detailed instructions on how to configure Alloy and add it to the `docker-compose.yaml` file are available in [our documentation](https://docs.openfaas.com/edge/open-telemetry/).
+
+Any instrumented function can be deployed to OpenFaaS Edge. Just make sure the `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable is set correctly.
+Modify the `stack.yaml` file and change the environment variable to `alloy:4317`. Or use the environment substitution to override it. In OpenFaaS Edge the service name can be used as the hostname.
+
+```sh
+# Use the stack.yaml from the previous section and override the OTLP endpoint.
+OTEL_EXPORTER_OTLP_ENDPOINT=alloy:4317 faas-cli up
+```
+
+## Conclusion
+
+In this tutorial we showed how OpenTelemetry zero-code instrumentation can be used to quickly get telemetry and traces for OpenFaaS functions without having to manually instrument your functions code.
```

Comment:
You should mention the Grafana solutions you used, and what other common alternatives exist i.e. [Jaeger and Zipkin](https://edgedelta.com/company/blog/the-battle-of-tracers-jaeger-vs-zipkin-the-complete-comparison).

For customers using Datadog, what do they need to do? Perhaps add a link - **https://docs.datadoghq.com/opentelemetry/**